### PR TITLE
Tags rework

### PR DESCRIPTION
**Pull request by JakaMohorkoDS**
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] Pull request title reflects its content
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description:

### CODE REVIEW

**Comment by JakaMohorkoDS** [comment]
@NikolaiShipilov Tags have been reworked to function similar to how the Component status works. The Component status should likely trigger the core events in a similar manner as the Tags, adding a new core event type.

File: core/opendaq/component/include/opendaq/component_impl.h
Comment (NikolaiShipilov) [inline core/opendaq/component/include/opendaq/component_impl.h:1]: The `tags` are serialized in `serializeCustomObjectValues(const SerializerPtr& serializer)`, doesn't it mean that `tags` should be also handled in `updateObject(const SerializedObjectPtr& obj)` ?

The similar question for #95 - should be the component statuses be also serialized / deserialized ?

File: core/opendaq/component/include/opendaq/tags_impl.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+    ErrCode INTERFACE_FUNC getList(IList** value) override;
+    ErrCode INTERFACE_FUNC add(IString* name) override;
+    ErrCode INTERFACE_FUNC remove(IString* name) override;
+    ErrCode INTERFACE_FUNC contains(IString* name, Bool* value) override;
+    ErrCode INTERFACE_FUNC query(IString* query, Bool* value) override;
+    
+    // IBaseObject
+
+    ErrCode INTERFACE_FUNC equals(IBaseObject* other, Bool* equals) const override;
```
Comment (janm-dw) [inline core/opendaq/component/include/opendaq/tags_impl.h:44]: If we have equals method implemented, then we should also think about implementing the getHashCode()

File: shared/libraries/opcuatms/opcuatms_client/src/objects/tms_client_tags_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    try
+    {
+        const ListPtr<IString> tagValues = VariantConverter<IString>::ToDaqList(client->readValue(nodeId));
+        this->tags.clear();
+        for (const auto& tag : tagValues)
+            this->tags.insert(tag);
+    }
+    catch([[maybe_unused]] const std::exception& e)
+    {
+        const auto loggerComponent = this->daqContext.getLogger().getOrAddComponent("OpcUaClient");
```
Comment (janm-dw) [inline shared/libraries/opcuatms/opcuatms_client/src/objects/tms_client_tags_impl.cpp:53]: Should loggerComponent be an instance variable*

**Review by dejancrn — APPROVED** [review]

File: core/opendaq/component/include/opendaq/tags_impl.h [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    if (tags.count(str) > 0)
+        return OPENDAQ_IGNORED;
+
+    tags.emplace(str);
+
+    if (triggerCoreEvent.assigned())
+    {
+        const auto thisPtr = this->borrowPtr<TagsPtr>();
+        const CoreEventArgsPtr args = createWithImplementation<ICoreEventArgs, CoreEventArgsImpl>(
+            core_event_ids::TagsChanged, Dict<IString, IBaseObject>({{"Tags", thisPtr}}));
```
[the remark this answers is not in the document]
Comment (JakaMohorkoDS) [reply core/opendaq/component/include/opendaq/tags_impl.h:101]: The tags object is always owned by a component, so I think it's fine without the path - the path is just the global ID of the sender component.

**Pull request merged**